Reorganize renderer code + cache and threading optimizations.

reduz opened the pull request:
Some benchmarks.
![image](https://user-images.githubusercontent.com/6265307/103646377-56536180-4f38-11eb-80e9-e859d20ce57e.png)

Dynamic Cull clearly needs a better system to do indexing.. the rest of the optimizations seem to have significantly improved performance. Still a good amount of room left to optimize.

reduz committed:
Reorganize renderer code.

So it can hopefully be made more cache efficient afterwards.

aaronfranke commented on servers/rendering/renderer_rd/renderer_scene_render_forward.h:
```diff
@@ -46,7 +47,8 @@ class RendererSceneRenderForward : public RendererSceneRenderRD {
 		SDFGI_MAX_CASCADES = 8,
-		MAX_GI_PROBES = 8
+		MAX_GI_PROBES = 8,
+		MAX_LIGHTMAPS = 8
```
Probably best to keep a trailing comma:

```suggestion
		MAX_LIGHTMAPS = 8,
```

reduz committed:
Rewrite render code to be more cache and thread friendly.

reduz committed:
Threaded optimizations to cull and render

-Reorganize thread work pool for rendering
-Fixes to make secondary command buffers to work (disabled because they need more testing)

akien-mga commented on servers/rendering/renderer_scene_cull.h:
```diff
@@ -325,7 +329,55 @@ class RendererSceneCull : public RendererScene {
+		bool mirror : 8;
+		bool receive_shadows : 8;
+		bool visible : 8;
+		bool baked_light : 2; //this flag is only to know if it actually did use baked light
+		bool dynamic_gi : 2; //this flag is only to know if it actually did use baked light
```
Copy paste sin :)

JFonS reviewed (approved):
I went over the files and I didn't see anything blatantly wrong, but it wasn't an in-depth review.
